tools/inundate_gms_optimized.py

```diff
@@ -0,0 +1,316 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+from concurrent.futures import ProcessPoolExecutor, as_completed
+
+import pandas as pd
+from inundation8 import NoForecastFound, hydroTableHasOnlyLakes, inundate
+from tqdm import tqdm
+
+from utils.shared_functions import FIM_Helpers as fh
+
+
+def Inundate_gms(
+    hydrofabric_dir,
+    forecast,
+    num_workers=1,
+    hydro_table_df=None,
+    hucs=None,
+    inundation_raster=None,
+    inundation_polygon=None,
+    depths_raster=None,
+    verbose=False,
+    log_file=None,
+    output_fileNames=None,
+):
+    # input handling
+    if hucs is not None:
+        try:
+            _ = (i for i in hucs)
+        except TypeError:
+            raise ValueError("hucs argument must be an iterable")
+
+    if isinstance(hucs, str):
+        hucs = [hucs]
+
+    num_workers = int(num_workers)
+
+    # log file
+    if log_file is not None:
+        if os.path.exists(log_file):
+            os.remove(log_file)
+
+        if verbose:
+            print("HUC8,BranchID,Exception", file=open(log_file, "w"))
+    # if log_file:
+    #     logging.basicConfig(filename=log_file, level=logging.INFO)
+    #     logging.info('HUC8,BranchID,Exception')
+
+    # load fim inputs
+    hucs_branches = pd.read_csv(
+        os.path.join(hydrofabric_dir, "fim_inputs.csv"), header=None, dtype={0: str, 1: str}
+    )
+
+    if hucs is not None:
+        hucs = set(hucs)
+        huc_indices = hucs_branches.loc[:, 0].isin(hucs)
+        hucs_branches = hucs_branches.loc[huc_indices, :]
+
+    # get number of branches
+    number_of_branches = len(hucs_branches)
+
+    # make inundate generator
+    inundate_input_generator = __inundate_gms_generator(
+        hucs_branches,
+        hydrofabric_dir,
+        inundation_raster,
+        inundation_polygon,
+        depths_raster,
+        forecast,
+        hydro_table_df,
+        verbose=False,
+    )
+
+    # start up process pool
+    # better results with Process pool
+    executor = ProcessPoolExecutor(max_workers=num_workers)
+
+    # collect output filenames
+    inundation_raster_fileNames = [None] * number_of_branches
+    inundation_polygon_fileNames = [None] * number_of_branches
+    depths_raster_fileNames = [None] * number_of_branches
+    hucCodes = [None] * number_of_branches
+    branch_ids = [None] * number_of_branches
+
+    executor_generator = {executor.submit(inundate, **inp): ids for inp, ids in inundate_input_generator}
+    idx = 0
+    for future in tqdm(
+        as_completed(executor_generator),
+        total=len(executor_generator),
+        desc=f"Inundating branches with {num_workers} workers",
+        disable=(not verbose),
+    ):
+        hucCode, branch_id = executor_generator[future]
+
+        try:
+            future.result()
+
+        except NoForecastFound as exc:
+            if log_file is not None:
+                print(f"{hucCode},{branch_id},{exc.__class__.__name__}, {exc}", file=open(log_file, "a"))
```

Comment:
opening a file here might not be a good idea. Python docs recommend explicitly closing opened files and not relying on garbage collection to properly flush and close the file.

Reply:
ya.. we use "with" for every io based write for log files.

Also.. just today, we went though a ton of our files looking more carefully if objects can be deleted earlier. We had a habit of loading a file and keeping it open the entire function/method even though we used it early on.  Much of FIM is now showing memory issues so we are cleaning those up. Good to keep an eye for it here too. You might already be doing it but just a heads up on it. :)